syntax/parser_arithm.go

```diff
@@ -0,0 +1,318 @@
+}
+
+func (p *Parser) arithmExpr3(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExpr2, Add, Sub)
+}
+
+func (p *Parser) arithmExpr2(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExprPower, Mul, Quo, Rem)
+}
+
+func (p *Parser) arithmExprPower(compact bool) ArithmExpr {
+	// Power is different from the other binary operators because it's right-associative
+	value := p.arithmExpr1(compact)
+	if BinAritOperator(p.tok) == Pow {
+		if compact && p.spaced {
+			return value
+		}
+		if value == nil {
+			p.curErr("%s must follow an expression", p.tok.String())
+		}
+		op := p.tok
+		pos := p.pos
+		p.nextArithOp(compact)
+		y := p.arithmExprPower(compact)
+		if y == nil {
+			p.followErrExp(pos, op.String())
+		}
+		return &BinaryArithm{
+			OpPos: pos,
+			Op:    BinAritOperator(op),
+			X:     value,
+			Y:     y,
+		}
+	}
+
+	return value
+}
+
+func (p *Parser) arithmExpr1(compact bool) ArithmExpr {
+	if !compact {
+		p.got(_Newl)
+	}
+
+	switch UnAritOperator(p.tok) {
+	case Not, BitNegation, Plus, Minus:
+		ue := &UnaryArithm{OpPos: p.pos, Op: UnAritOperator(p.tok)}
+		p.nextArithOp(compact)
+		if ue.X = p.arithmExpr1(compact); ue.X == nil {
+			p.followErrExp(ue.OpPos, ue.Op.String())
+		}
+		return ue
+	}
+
+	return p.arithmExpr0(compact)
+}
+
+func (p *Parser) arithmExpr0(compact bool) ArithmExpr {
+	var x ArithmExpr
+	switch p.tok {
+	case addAdd, subSub:
+		ue := &UnaryArithm{OpPos: p.pos, Op: UnAritOperator(p.tok)}
+		p.nextArith(compact)
+		if p.tok != _LitWord {
+			p.followErr(ue.OpPos, token(ue.Op).String(), "a literal")
+		}
+		ue.X = p.arithmExpr0(compact)
+		return ue
+	case leftParen:
+		pe := &ParenArithm{Lparen: p.pos}
+		p.nextArithOp(compact)
+		pe.X = p.followArithm(leftParen, pe.Lparen)
+		pe.Rparen = p.matched(pe.Lparen, leftParen, rightParen)
+		x = pe
+	case _LitWord:
+		l := p.getLit()
+		if p.tok != leftBrack {
+			x = p.word(p.wps(l))
+			break
+		}
+		pe := &ParamExp{Dollar: l.ValuePos, Short: true, Param: l}
+		pe.Index = p.eitherIndex()
+		x = p.word(p.wps(pe))
+	case bckQuote:
+		if p.quote == arithmExprLet && p.openBquotes > 0 {
+			return nil
+		}
+		fallthrough
+	default:
+		if w := p.getWord(); w != nil {
+			x = w
+		} else {
+			return nil
+		}
+	}
+	if compact && p.spaced {
+		return x
+	}
+	for !compact && p.got(_Newl) {
+	}
+
+	// we want real nil, not (*Word)(nil) as that
+	// sets the type to non-nil and then x != nil
+	if p.tok == addAdd || p.tok == subSub {
+		if !isArithName(x) {
+			p.curErr("%s must follow a name", p.tok.String())
+		}
+		u := &UnaryArithm{
+			Post:  true,
+			OpPos: p.pos,
+			Op:    UnAritOperator(p.tok),
+			X:     x,
+		}
+		p.nextArith(compact)
+		return u
+	}
+
+	return x
+}
+
+// nextArith consumes a token.
+// It returns true if compact and the token was followed by spaces
+func (p *Parser) nextArith(compact bool) bool {
+	p.next()
+	if compact && p.spaced {
+		return true
+	}
+	for !compact && p.got(_Newl) {
+	}
```

Comment:
if I recall correctly, you should only care about a single _Newl token, since they are deduplicated. so you could do:

```
if !compact {
    p.got(_Newl)
}
```

More lines, but also simpler, and it's best to avoid loops in general as they just add more ways for the parser to get stuck or burn cpu forever.

Reply:
ah didn't know they're deduplicated by the lexer, will do